### DISCUSSION TITLE
Access dependencies file in python

### DISCUSSION THREAD

**Discussion by dflvunoooooo** [leo-colisson/robust-externalize#27] [Q&A] [answered]
Is it possible to access a file specified with the dependencies option inside a CacheMeCode environment? This way the filename would only have to be specified once. Furthermore does python require one to explicit look in the parent dircectory if the imported file is not in the robustExternalize folder. And that is not adviced, see here [#10](https://github.com/leo-colisson/robust-externalize/issues/10#issuecomment-1855650941). 

Now one has to do this:
```
\begin{CacheMeCode}{python, dependencies={name.dat}}
…
file = '../name.dat'
```

**Comment by tobiasBora** [comment, marked as answer]
One thing you can do is to use a placeholder + create a style like:
```
\robExtConfigure{
  add deps placeholder/.style={
    dependencies = {#1},
    set placeholder={__MY_DEPS__}{__ROBEXT_WAY_BACK__#1},
  },
}
\begin{CacheMeCode}{python, add deps placeholder={name.dat}}
…
file = '__MY_DEPS__'
```
This way the `../` is automatically added. If you find is useful, I can consider to add a similar functionality.

In general, I don't know a generic equivalent of the loadpath for python, but I guess the above code should make the `../` invisible enough not to be a problem, right? You can also use `copy file to cache`, but it is less elegant as it will copy the file to the cache which might be an issue if the file is really large.

**Reply by dflvunoooooo** [reply]
Thank you for your answer. That will help thanks. There is no way to access the dependencies from inside python?

**Reply by tobiasBora** [reply]
What do you mean? In the above example `__MY_DEPS__` will be readable by python directly.

**Reply by dflvunoooooo** [reply]
But is it the path to the file, too? I mean something like that
```
\begin{CacheMeCode}{python, dependencies={folder/name.dat}}
…
file = '__ROBEXT_DEPENDENCY_1__'
```
So that I only have to define the path once. From inside Python it would otherwise be `'../folder/name.dat'`.

**Reply by tobiasBora** [reply]
yes exactly, with my above example `__MY_DEPS__` will point to the file `../name.dat` so that you just need to type it once. If it fails, just provide a MWE that I can try.

**Reply by dflvunoooooo** [reply]
It works, if I do not misspell the commands. Thank you very much! :)

**Reply by tobiasBora** [reply]
Great!

**Reply by dflvunoooooo** [reply]
Can there be more than one such placeholder, if I want to import more than one file? Adding another `add deps placeholder={folder/name2.dat}` does not work or I don't know how to access this placeholder. I can not give names to `
add deps placeholder`, at least it did not work just now.

**Reply by tobiasBora** [reply]
Sure! The simplest is to use `add dependencies` instead, and create multiple styles like in:
```
\robExtConfigure{
  add deps placeholder 1/.style={
    add dependencies = {#1},
    set placeholder={__MY_DEPS_1__}{__ROBEXT_WAY_BACK__#1},
  },
  add deps placeholder 2/.style={
    add dependencies = {#1},
    set placeholder={__MY_DEPS_2__}{__ROBEXT_WAY_BACK__#1},
  },
  add deps placeholder 3/.style={
    add dependencies = {#1},
    set placeholder={__MY_DEPS_3__}{__ROBEXT_WAY_BACK__#1},
  },
}
\begin{CacheMeCode}{python, add deps placeholder 1={name1.dat}, add deps placeholder 2={name2.dat}, add deps placeholder 3={name3.dat}}
…
fileA = '__MY_DEPS_1__'
fileB = '__MY_DEPS_2__'
fileC = '__MY_DEPS_2__'
```
You can also create a python list, but it requires more non-trivial coding.

**Reply by tobiasBora** [reply]
Also, you can create named placeholders like this (might be cleaner and more generic actually):
```
\robExtConfigure{
  add named deps placeholder/.style 2 args={
    add dependencies = {#2},
    set placeholder={__MY_DEPS_#1__}{__ROBEXT_WAY_BACK__#2},
  },
}
\begin{CacheMeCode}{python, add named deps placeholder={FIRST}{name1.dat}, add named deps placeholder={SECOND}{name2.dat}}
…
fileA = '__MY_DEPS_FIRST__'
fileB = '__MY_DEPS_SECOND__'
```

**Reply by dflvunoooooo** [reply]
> Sure! The simplest is to use `add dependencies` instead, and create multiple styles like in:
…
> fileA = '__MY_DEPS_1__'
> fileB = '__MY_DEPS_2__'
> fileC = '__MY_DEPS_2__'
> ```
> 
> You can also create a python list, but it requires more non-trivial coding.

Perfekt. Thank you very much! This is fine for me!

**Reply by dflvunoooooo** [reply]
> Also, you can create named placeholders like this (might be cleaner and more generic actually):
> 
> ```
> \robExtConfigure{
>   add named deps placeholder/.style 2 args={
>     add dependencies = {#2},
>     set placeholder={__MY_DEPS_#1__}{__ROBEXT_WAY_BACK__#2},
>   },
> }
> \begin{CacheMeCode}{python, add named deps placeholder={FIRST}{name1.dat}, add named deps placeholder={SECOND}{name2.dat}}
> …
> fileA = '__MY_DEPS_FIRST__'
> fileB = '__MY_DEPS_SECOND__'
> ```

More generic, this looks better. Thanks alot!
